src/transformers/cache_utils.py

```diff
@@ -320,3 +320,361 @@ def reorder_cache(self, beam_idx: torch.LongTensor):
+                device=value_states.device,
+            )
+            for batch_idx in range(batch_size):
+                seq_id = self.batch2seq[batch_idx][0]
+                for i in range(new_context_len):
+                    block_idx = self.block_tables[seq_id][i // self.block_size]
+                    block_offset = i % self.block_size
+                    key[batch_idx][i] = self.key_cache[layer_idx][block_idx][
+                        block_offset
+                    ]
+                    value[batch_idx][i] = self.value_cache[layer_idx][block_idx][
+                        block_offset
+                    ]
+            key_states = key.transpose(1, 2).contiguous()
+            value_states = value.transpose(1, 2).contiguous()
+        return key_states, value_states
+
+    def update(
+        self,
+        key_states: torch.Tensor,
+        value_states: torch.Tensor,
+        layer_idx: int,
+        cache_kwargs: Optional[Dict[str, Any]] = None,
+    ) -> Tuple[torch.Tensor, torch.Tensor]:
+        """
+        Update the cache with key and value states for a specific layer.
+
+        Args:
+            key_states (torch.Tensor): The key states tensor of shape [batch, head, seq, dim].
+            value_states (torch.Tensor): The value states tensor of shape [batch, head, seq, dim].
+            layer_idx (int): The index of the layer.
+            cache_kwargs (Dict[str, Any]): Additional arguments for the cache subclass.
+        Returns:
+            Tuple[torch.Tensor, torch.Tensor]: A tuple containing the updated key states and value states tensors.
+
+        Raises:
+            AssertionError: If the batch size is inconsistent with the existing cache.
+        """
+        batch_size = key_states.shape[0]  # [batch, head, seq, dim]
+        kv_head = key_states.shape[1]
+        head_size = key_states.shape[-1]
+        if layer_idx == 0:
+            self.seen_tokens += key_states.shape[-2]
+        original_key_states = key_states
+        # self.batch2seq is only for the current decode step, need to clear in the last layer and init in the first layer or setup externally
+        if layer_idx == 0 and self.batch2seq == {}:
+            assert len(self.block_tables) == 0
+            self.batch2seq = {i: [i] for i in range(batch_size)}
+            self.slots_mapping = []
+        elif layer_idx == 0 and self.batch2seq != {}:
+            assert len(self.batch2seq) == batch_size
+
+        if layer_idx not in self.key_cache:  # init the cache
+            self.key_cache[layer_idx] = torch.zeros(
+                (self.num_blocks, self.block_size, kv_head, head_size),
+                dtype=key_states.dtype,
+                device=key_states.device,
+            )
+            self.value_cache[layer_idx] = torch.zeros(
+                (self.num_blocks, self.block_size, kv_head, head_size),
+                dtype=value_states.dtype,
+                device=value_states.device,
+            )
```

Comment:
As I see it, this initializes the whole cache at once, as opposed to allocating new slots as needed 🤔 Not allocating everything at once is one of the major selling points of `PagedAttention`, otherwise it is a complicated version of a static cache (which we are working on on a separate PR), but with the ability sharing past KV for duplicate prompts.

Supporting dynamically growing caches would be nice, but it might be tricky to implement. If it proves to be hard/a performance killer, then we should rename the class, as it wouldn't be `PagedAttention`

Reply:
Per my understand, the core idea of pagedAttention as following:

- The _block_ is the least allocation unit in paged attention and there is _block_size_ slots in a block. 
- 1 sequence can be divided into multiple logical blocks. Token intra-block is contiguously stored, and contiguous logic block can be stored in non-contiguous physical blocks which is the real memory space. So, need a mapping table to map the logical block to physical block. 
- PagedAttention consume less memory space than native _kv_cache_. The wasted memory space is no more than 1 physical block for every sequence with pagedAttention.  But for the native kv_cache,  all sentences need to be padded to the _max_seq_len_ which will waste a lot of memory space when the real sequence length is less than the _max_seq_len_ .  
![image](https://github.com/huggingface/transformers/assets/46986936/496b27f9-8d4d-4ab0-952c-808f299cbd17)

Allocate whole cache at once is not conflict with the core idea of  pagedAttention and the naive implementation in vLLM also do this in the same way. pls refer to this link  https://github.com/vllm-project/vllm/blob/main/vllm/worker/cache_engine.py#L70

Reply:
I see: it allocates all memory at once, but only *consumes* as needed. Gotcha 👍 

(When I read the paper, I read it as "we allocate as we need", and they tracked allocated memory. Allocating all at once is much more efficient, though -- they allocate all at once, and track consumed memory.)